apps/src/turtle/artist.js

```diff
@@ -726,23 +724,26 @@ Artist.prototype.resetButtonClick = function () {
   }
 };
 
-Artist.prototype.evalCode = function (code) {
+Artist.prototype.evalCode = function (code, interpreterScope={}) {
   try {
     CustomMarshalingInterpreter.evalWith(code, {
-      Turtle: this.api
+      Turtle: this.api,
+      ...interpreterScope,
     });
   } catch (e) {
     // Infinity is thrown if we detect an infinite loop. In that case we'll
     // stop further execution, animate what occurred before the infinite loop,
     // and analyze success/failure based on what was drawn.
     // Otherwise, abnormal termination is a user error.
-    if (e !== 'Infinity') {
+    if (e.message !== 'Infinity') {
```

Comment:
I vaguely remember fixing this up when we switched over to running in the JS Interpreter (commit https://github.com/code-dot-org/code-dot-org/commit/6cb471716825dd262b85a31072200c02166206d1).  There must've been some reason I changed it to the string, instead of checking the message?  Or maybe a more recent interpreter update made this moot — but then shouldn't a unit test have caught it?

Anyway, maybe worth double checking we don't throw the string `'Infinity'` somewhere, to handle both cases.

Reply:
The error that we were missing here *was* being thrown as the string `'Infinity'`, but the JSInterpreter wraps non-Error objects that get thrown by native functions in an Error object: https://github.com/code-dot-org/code-dot-org/blob/6891ee0f53254df60f7cde3183449f27b85bd371/apps/src/lib/tools/jsinterpreter/CustomMarshalingInterpreter.js#L790-L794.

There's an infinite loop unit test, but it catches a different error thrown here: https://github.com/code-dot-org/blockly/blob/b4e845c4f83326af3907e82cc5cd00fddc445f54/generators/javascript/loops.js#L120. Since that is thrown from interpreted code instead of a native function, it was handled differently by the interpreter.

I'm changing both errors to `new Error('Infinity')`, which the interpreter handles in a consistent fashion.

Reply:
Nice, thanks for tracking that down!